src/modules/mc_pos_control/MulticopterPositionControl.cpp

```diff
@@ -336,36 +348,66 @@ void MulticopterPositionControl::Run()
 			}
 		}
 
-		PositionControlStates states{set_vehicle_states(local_pos)};
+		const bool is_trajectory_setpoint_updated = _trajectory_setpoint_sub.update(&_setpoint);
 
-		if (_vehicle_control_mode.flag_multicopter_position_control_enabled) {
+		// adjust existing (or older) setpoint with any EKF reset deltas
+		if ((_setpoint.timestamp != 0) && (_setpoint.timestamp < local_pos.timestamp)) {
+			if (local_pos.vxy_reset_counter != _vxy_reset_counter) {
+				_setpoint.vx += local_pos.delta_vxy[0];
+				_setpoint.vy += local_pos.delta_vxy[1];
+			}
 
-			const bool is_trajectory_setpoint_updated = _trajectory_setpoint_sub.update(&_setpoint);
+			if (local_pos.vz_reset_counter != _vz_reset_counter) {
+				_setpoint.vz += local_pos.delta_vz;
+			}
 
-			// adjust existing (or older) setpoint with any EKF reset deltas
-			if (_setpoint.timestamp < local_pos.timestamp) {
-				if (local_pos.vxy_reset_counter != _vxy_reset_counter) {
-					_setpoint.vx += local_pos.delta_vxy[0];
-					_setpoint.vy += local_pos.delta_vxy[1];
-				}
+			if (local_pos.xy_reset_counter != _xy_reset_counter) {
+				_setpoint.x += local_pos.delta_xy[0];
+				_setpoint.y += local_pos.delta_xy[1];
+			}
 
-				if (local_pos.vz_reset_counter != _vz_reset_counter) {
-					_setpoint.vz += local_pos.delta_vz;
-				}
+			if (local_pos.z_reset_counter != _z_reset_counter) {
+				_setpoint.z += local_pos.delta_z;
+			}
 
-				if (local_pos.xy_reset_counter != _xy_reset_counter) {
-					_setpoint.x += local_pos.delta_xy[0];
-					_setpoint.y += local_pos.delta_xy[1];
-				}
+			if (local_pos.heading_reset_counter != _heading_reset_counter) {
+				_setpoint.yaw += local_pos.delta_heading;
+			}
+		}
 
-				if (local_pos.z_reset_counter != _z_reset_counter) {
-					_setpoint.z += local_pos.delta_z;
-				}
+		if (local_pos.vxy_reset_counter != _vxy_reset_counter) {
+			_vel_x_deriv.reset();
+			_vel_y_deriv.reset();
+		}
 
-				if (local_pos.heading_reset_counter != _heading_reset_counter) {
-					_setpoint.yaw += local_pos.delta_heading;
-				}
+		if (local_pos.vz_reset_counter != _vz_reset_counter) {
+			_vel_z_deriv.reset();
+		}
+
+		// save latest reset counters
+		_vxy_reset_counter = local_pos.vxy_reset_counter;
+		_vz_reset_counter = local_pos.vz_reset_counter;
+		_xy_reset_counter = local_pos.xy_reset_counter;
+		_z_reset_counter = local_pos.z_reset_counter;
+		_heading_reset_counter = local_pos.heading_reset_counter;
+
+
+		PositionControlStates states{set_vehicle_states(local_pos)};
+
+
+		if (_vehicle_control_mode.flag_multicopter_position_control_enabled) {
+			// set failsafe setpoint if there hasn't been a new
+			// trajectory setpoint since position control started
+			if ((_setpoint.timestamp < _time_position_control_enabled)
+			    && (time_stamp_now > _time_position_control_enabled)) {
```

Comment:
It actually provides a slight time delay because `time_stamp_now = vehicle_local_position.timestamp_sample`.

We could make it more explicit.

Reply:
> It actually provides a slight time delay because

I made it more explicit what the line checks but I honestly still don't understand what it's useful for or how it provides a delay. The first condition intuitively says if the setpoint is still from before position control was enabled we brige it with the failsafe setpoint. But why would you not do that and try to execute the old, reset, invalid setpoint in case the current local position sample is older than when position control was enabled?

Reply:
So combining with the line setting the timestamp when position control was enabled above:
`_time_position_control_enabled = _vehicle_control_mode.timestamp;`
the condition
`vehicle_local_position.timestamp_sample > _time_position_control_enabled`
does not bridge the old, reset, invalid setpoint on the first sample position control is enabled.

Shouldn't we remove that?